File: pdl/infra/java/src/main/java/io/pdl/infrastructure/data/QueryBuilder.java
package io.pdl.infrastructure.data;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public final class QueryBuilder {
    private final String table;
    private DBStore store;
    private final List<Filter> filters;
    private final List<String> projection;
    private final List<OrderClause> orderings;
    private Integer limit;
    private Integer offset;

    public QueryBuilder(String table) {
        this(table, null);
    }

    public QueryBuilder(String table, DBStore store) {
        this.table = table;
        this.store = store;
        this.filters = new ArrayList<>();
        this.projection = new ArrayList<>();
        this.orderings = new ArrayList<>();
    }

    public static QueryBuilder table(String table) {
        return new QueryBuilder(table);
    }

    public QueryBuilder withStore(DBStore store) {
        this.store = store;
        return this;
    }

    public QueryBuilder eq(String field, Object value) {
        return filter(field, Operator.EQ, value);
    }

    public QueryBuilder neq(String field, Object value) {
        return filter(field, Operator.NEQ, value);
    }

    public QueryBuilder in(String field, Object value) {
        return filter(field, Operator.IN, value);
    }

    public QueryBuilder filter(String field, Operator operator, Object value) {
        filters.add(new Filter(field, operator, value));
        return this;
    }

    public QueryBuilder project(String... columns) {
        Collections.addAll(projection, columns);
        return this;
    }

    public QueryBuilder orderBy(String column, OrderDirection direction) {
        if (column == null || column.isEmpty()) {
            return this;
        }
        orderings.add(new OrderClause(column, direction));
        return this;
    }

    public QueryBuilder asc(String column) {
        return orderBy(column, OrderDirection.ASC);
    }

    public QueryBuilder desc(String column) {
        return orderBy(column, OrderDirection.DESC);
    }

    public QueryBuilder limit(int value) {
        this.limit = value;
        return this;
    }

    public QueryBuilder offset(int value) {
        this.offset = value;
        return this;
    }

    public QueryBuilder range(int offsetValue, int limitValue) {
        this.offset = offsetValue;
        this.limit = limitValue;
        return this;
    }

    public List<Map<String, Object>> load() throws Exception {
        DBStore resolved = Row.resolveStore(store);
        return resolved.select(table, filters, projection, orderings, limit, offset);
    }

    public void delete(String primaryKey) throws Exception {
        DBStore resolved = Row.resolveStore(store);
        List<String> primaryProjection = new ArrayList<>(1);
        primaryProjection.add(primaryKey);
        List<Map<String, Object>> rows = resolved.select(table, filters, primaryProjection, orderings, limit, offset);
        for (Map<String, Object> entry : rows) {
            resolved.delete(table, primaryKey, entry);
        }
    }
}

File: pdl/infra/csharp/Io/Pdl/Infrastructure/Data/QueryBuilder.cs
using System.Collections.Generic;
using System.Linq;

namespace Io.Pdl.Infrastructure.Data;

public sealed class QueryBuilder
{
    private readonly List<Filter> _filters = new();
    private readonly List<string> _projection = new();
    private readonly List<OrderClause> _orderings = new();
    private int? _limit;
    private int? _offset;
    private IDbStore? _store;

    public QueryBuilder(string table, IDbStore? store)
    {
        Table = table;
        _store = store;
    }

    public string Table { get; }

    public QueryBuilder WithStore(IDbStore? store)
    {
        _store = store;
        return this;
    }

    public QueryBuilder Filter(string column, Operator op, object? value)
    {
        _filters.Add(new Filter(column, op, value));
        return this;
    }

    public QueryBuilder Project(params string[] columns)
    {
        _projection.AddRange(columns);
        return this;
    }

    public QueryBuilder Offset(int value)
    {
        _offset = value;
        return this;
    }

    public QueryBuilder Limit(int value)
    {
        _limit = value;
        return this;
    }

    public QueryBuilder Range(int offset, int limit)
    {
        _offset = offset;
        _limit = limit;
        return this;
    }

    public QueryBuilder OrderBy(string column, OrderDirection direction)
    {
        _orderings.Add(new OrderClause(column, direction));
        return this;
    }

    public IReadOnlyList<IDictionary<string, object?>> Load()
    {
        var store = StoreRegistry.Resolve(_store);
        return store.Select(Table, _filters, _projection, _orderings, _limit, _offset);
    }

    public void Delete(string primaryKey)
    {
        var store = StoreRegistry.Resolve(_store);
        var rows = store.Select(Table, _filters, new[] { primaryKey }, new List<OrderClause>(), null, null);
        foreach (var row in rows)
        {
            store.Delete(Table, primaryKey, row);
        }
    }
}
